docs/helpers.md

```diff
@@ -0,0 +1,97 @@
+Helper Structs
+--------------
+
+##### Memo
+
+Linked list, optimized for quickly finding a given type.
+
+- type: int, either a token or a rule (rules start at 256)
+- node: NULL or pointer to AST node OR pointer to token object
+- mark: if node != NULL, index into Parser's array of tokens
+- next: NULL or pointer to next Memo structure
+
+##### Token
+
+These are in an array linked from Parser.  (Or linked list???)
+
+- type: int, token type (needs only 8 bits)
+- bytes: bytes object
+- lineno, col_offset, end_lineno, end_col_offset: int
+- memo: Pointer to linked list Memo
+
+##### Parser
+
+Also needs to point to a PyArena, used for allocating AST nodes and
+other things.  And maybe some flag indicating there's an allocation
+error and a convention to bail if there is one.
+
+- tok: Pointer to tokenizer, CPython's struct tok_state
+- tokens: Pointer to array of Token pointers
+- mark: index into array of Tokens
+- fill: number of valid entries in array of Tokens
+- size: total number of entries in array of Tokens
+- arena: memory allocation arena (owns all AST, Token, Memo structures allocated)
+
+##### PegenAlias
+
+Needed because alias_ty does not hold line and column offset info.
+When a rule needs to return an alias_ty, it returns a PegenAlias*
+instead so that line and column info get propagated to the caller,
+which extracts this info together with the alias_ty.
+
+- alias: alias_ty
+- lineno, col_offset, end_lineno, end_col_offset: int
+
+##### CmpopExprPair
+
+This gets used by the rules that implement comparison, due to the
+structure of the Compare AST Object.
+
+- cmpop: cmpop_ty, The comparison operator
+- expr: expr_ty, The expression that gets compared
+
+
+Helper Functions
+----------------
+
+###### `asdl_seq *singleton_seq(Parser *, void *);`
+Creates a single-element asdl_seq* that contains a
```

Comment:
That contains a what? :-) I think you forgot to add the argument names to the header, and I recommend putting `asdl_seq*` and `a` in backticks. Also add a period.

Reply:
I'll do that. I also think that the semicolons should be removed.